aten/src/ATen/Quantizer.h

```diff
@@ -0,0 +1,202 @@
+#pragma once
+
+#include <c10/core/QScheme.h>
+#include <c10/macros/Macros.h>
+#include <c10/util/Exception.h>
+#include <ATen/core/Tensor.h>
+
+#include <cmath>
+#include <memory>
+
+// TODO: move to c10 namespace after we
+// unified caffe2::Tensor and at::Tensor
+
+namespace at {
+
+using QTensor = Tensor;
+using RealTensor = Tensor;
+
+/**
+ * Quantizer is the class for storing all the information
+ * that's necessary to perform quantize and dequantize
+ * operation.
+ *
+ * We might have different types of quantization schemes and this is
+ * the base class for all quantizers.
+ *
+ * QTensorImpl will hold a pointer to Quantizer so that we can support
+ * different quantization schemes on Tensor.
+ *
+ * For example, the most common quantization scheme, Affine Quantization, requires
+ * scale and zero_point as parameters, we'll store scale and zero_point inside
+ * the instance and we can use it to quantize a float Tensor or dequantize a
+ * quantized Tensor.
+ */
+struct C10_API Quantizer {
+  c10::QScheme qscheme_;
+  virtual ~Quantizer() {}
+
+  virtual std::string name() {
+    return "Quantizer";
+  }
+
+  /**
+   * quantize a float Tensor into a quantized Tensor.
+   */
+  virtual QTensor quantize(RealTensor t) = 0;
+
+  /**
+   * dequantize a quantized Tensor into a float Tensor.
+   */
+  virtual RealTensor dequantize(QTensor t) = 0;
+};
+
+/**
+ * UniformQuantizer is the parent class for all uniform quantizers.
+ * These quantization scheme will map float value uniformly to
+ * the quantized value. For example, affine quantizer is
+ * the most commonly used scheme in this category.
+ */
+struct C10_API UniformQuantizer : public Quantizer {
+  virtual ~UniformQuantizer() {}
+  virtual std::string name() {
+    return "UniformQuantizer";
+  }
+};
+
+/**
+ * NonUniformQuantizer is the parent class for all non-uniform quantizers.
+ * These quantization scheme may map float value non-uniformly to the quantized
+ * value. K-means quantization is a representative example in this category.
+ */
+struct C10_API NonUniformQuantizer : public Quantizer {
+  virtual ~NonUniformQuantizer() {}
+
+  virtual std::string name() {
+    return "NonUniformQuantizer";
+  }
+};
+
+// There is also StochasticQuantizer which is uniform but not affine
+
+/**
+ * AffineQuantizer uses affine transformation to do quantization.
+ *
+ * For quantize:
+ * Y = clamp((X * scale + zero_point, min, max)
+ * For dequantize:
+ * X = (Y - zero_point) / scale
+ */
+struct C10_API AffineQuantizer : public UniformQuantizer {
+  virtual ~AffineQuantizer() {}
+
+  virtual std::string name() {
+    return "AffineQuantizer";
+  }
```

Comment:
So.... these classes are not complete?

Reply:
Yeah only the leaf classes are actual quantizers